core/src/blocktree.rs

```diff
@@ -453,6 +579,79 @@ impl Blocktree {
         self.erasure_cf.put_bytes((slot, index), bytes)
     }
 
+    /// this function will insert coding blobs and also automatically track erasure-related
+    /// metadata
+    /// The `auto_recover` parameter controls whether blocktree will automatically do any recovery
+    /// made possible by the insertion of this coding blob
+    #[cfg(feature = "erasure")]
+    pub fn write_coding_blob(
+        &self,
+        slot: u64,
+        index: u64,
+        bytes: &[u8],
+        auto_recover: bool,
+    ) -> Result<()> {
+        let set_index = ErasureMeta::set_index_for(index);
+        let mut erasure_meta = self
+            .erasure_meta_cf
+            .get((slot, set_index))?
+            .unwrap_or_else(|| ErasureMeta::new(slot, set_index));
+
+        erasure_meta.set_coding_present(index);
+
+        let mut writebatch = self.db.batch()?;
+
+        writebatch.put_bytes::<cf::Coding>((slot, index), bytes)?;
+
+        let mut blobs_to_write = vec![];
+        let mut coding_to_delete = vec![];
+
+        if auto_recover && erasure_meta.can_recover() {
+            match self.recover(&erasure_meta, slot, index, bytes) {
+                Ok((mut recovered_data, recovered_coding)) => {
+                    blobs_to_write.append(&mut recovered_data);
+
+                    if !recovered_coding.is_empty() {
+                        for coding_blob in recovered_coding {
+                            let blob = coding_blob.read().unwrap();
+                            let index = blob.index();
+
+                            writebatch.put_bytes::<cf::Coding>(
+                                (slot, index),
+                                &blob.data[..BLOB_HEADER_SIZE + blob.size()],
+                            )?;
+
+                            erasure_meta.set_coding_present(index)
+                        }
+                    }
+                }
+                Err(Error::ErasureError(erasure::ErasureError::CorruptCoding)) => {
+                    let (_, coding_start) = erasure_meta.start_indices();
+                    let coding_end = coding_start + erasure::NUM_DATA as u64;
+
+                    erasure_meta.coding = 0;
+
+                    for i in coding_start..coding_end {
```

Comment:
with a slight re-jiggering of broadcast to index the coding blobs starting at zero, a lot of this work goes away

coding blob indexing being from [12-16) is an accident of history, partly the way the window used to work, partly the one-entry-per-blob restrictions

Reply:
Should I do that in this PR or separate that work out?

Reply:
separating it out might end up being as big as this PR